ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java

```diff
@@ -208,44 +207,55 @@ public void resourceChanged(IResourceChangeEvent event) {
 		IResourceDelta delta = event.getDelta();
 
 		try {
+			List<IFile> added = new ArrayList<>();
+			List<IFile> removed = new ArrayList<>();
 			delta.accept(delta1 -> {
-				int kind = delta1.getKind();
 				IResource resource = delta1.getResource();
-				if (resource instanceof IFile) {
+				if (monitoredFiles.contains(resource)) {
```

Comment:
This is almost perfect, just we do not need to perform 100_000 times lookup in a map if the delta isn't interesting for us (which is true in most cases).
So consider this slight modification:
```
diff --git a/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java b/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java
index c14d606..55e5538 100644
--- a/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java
+++ b/ui/org.eclipse.pde.ui/src/org/eclipse/pde/internal/ui/editor/context/InputContextManager.java
@@ -212,9 +212,13 @@
 			delta.accept(delta1 -> {
+				int kind = delta1.getKind();
+				if (kind != IResourceDelta.ADDED && kind != IResourceDelta.REMOVED) {
+					return true;
+				}
 				IResource resource = delta1.getResource();
 				if (monitoredFiles.contains(resource)) {
-					int kind = delta1.getKind();
-					if (kind == IResourceDelta.ADDED)
+					if (kind == IResourceDelta.ADDED) {
 						added.add((IFile) resource);
-					else if (kind == IResourceDelta.REMOVED)
+					} else if (kind == IResourceDelta.REMOVED) {
 						removed.add((IFile) resource);
+					}
 					return false;

```